src/txn/txn_rollback_to_stable.c

```diff
@@ -53,12 +51,14 @@ __rollback_delete_hs(WT_SESSION_IMPL *session, WT_ITEM *key, wt_timestamp_t ts)
     hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
     ret = __wt_curhs_search_near_before(session, hs_cursor);
     for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
-        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
-        if (hs_start_ts < ts)
+        /* Retrieve the time window from the history cursor. */
+        __wt_hs_upd_time_window(hs_cursor, &hs_tw);
+        if (hs_tw->start_ts < ts)
             break;
+
```

Comment:
I think it is ok to break. If U2 and U1 are already in the history store, they should be with 0 timestamp already.

Reply:
I suppose it's not possible to read U2 after RTS because it's not possible for U3 to be invisible at that point. (Since you can't call RTS with a transaction active, it shouldn't be possible to read afterwards with a snapshot from before, so U3 will always be visible.)

In that case though maybe RTS should remove both U1 and U2... hmm. Not now I guess.